sites/main-site/src/content/blog/2024/seqera-containers-part-2.mdx

```diff
@@ -0,0 +1,623 @@
+
+With [the FastQC module](https://github.com/nf-core/modules/blob/f768b283dbd8fc79d0d92b0f68665d7bed94cabc/modules/nf-core/fastqc/main.nf#L6-L8)
+as an example:
+
+```diff title="main.nf"
+process FASTQC {
+     label 'process_medium'
+
+     conda "${moduleDir}/environment.yml"
++    container "fastqc:0.12.1--5cfd0f3cb6760c42" // automatically generated
+-    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+-        'https://depot.galaxyproject.org/singularity/fastqc:0.12.1--hdfd78af_0' :
+-        'biocontainers/fastqc:0.12.1--hdfd78af_0' }"
+
+     input:
+     tuple val(meta), path(reads)
+```
+
+We considered removing both `conda` and `container` declarations from the module `main.nf` file entirely,
+however we see benefit for keeping these in this form because:
+
+-   It's clearer to those exploring the code about what the module requires.
+-   The container string is needed to tie the module to the pipeline config files
+    (see [Building config files](#building-config-files) below).
+
+Removing the container logic from the string should be a big win for readability.
+
+### `meta.yml`
+
+Through the magic of automation, we will append and then validate the following fields
+within the module's `meta.yml` file. Following the
+[FastQC example](https://github.com/nf-core/modules/blob/f768b283dbd8fc79d0d92b0f68665d7bed94cabc/modules/nf-core/fastqc/meta.yml)
+from above:
+
+```yaml title="meta.yml"
+# ..existing meta.yml content above
+containers:
+    docker:
+        linux_amd64:
+            name: community.wave.seqera.io/library/fastqc:0.12.1--5cfd0f3cb6760c42
+            build_id: 5cfd0f3cb6760c42_1
+            scan_id: 6fc310277b74
+        linux_arm64:
+            name: community.wave.seqera.io/library/fastqc:0.12.1--d3caca66b4f3d3b0
+            build_id: d3caca66b4f3d3b0_1
+            scan_id: d9a1db848b9b
+    singularity:
+        linux_amd64:
+            name: oras://community.wave.seqera.io/library/fastqc:0.12.1--0827550dd72a3745
+            https: https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/b2/b280a35770a70ed67008c1d6b6db118409bc3adbb3a98edcd55991189e5116f6/data
+            build_id: 0827550dd72a3745_1
+        linux_arm64:
+            name: oras://community.wave.seqera.io/library/fastqc:0.12.1--b2ccdee5305e5859
+            https: https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/76/76e744b425a6b4c7eb8f12e03fa15daf7054de36557d2f0c4eb53ad952f9b0e3/data
+            build_id: b2ccdee5305e5859_1
+    conda:
+        linux_amd64:
+            lockfile: https://wave.seqera.io/v1alpha1/builds/5cfd0f3cb6760c42_1/condalock
+        linux_arm64:
+            lockfile: https://wave.seqera.io/v1alpha1/builds/d3caca66b4f3d3b0_1/condalock
+```
+
+The four different container images are all built at the same time so share the same snapshot of a given Conda resolution.
```

Comment:
What does it mean "the same snapshot of a given Conda resolution"? 
It means that they will ping the same version of the Conda packages but are built independently or there is some more magic here that I am not aware?

Reply:
Yeah this is poorly written. I just mean that they are built at the same point in time, so it's not like building them six months apart when something in the dependency tree has changed. I'll update.

> All images are all built at the same time, avoiding Conda dependency drift.

Reply:
Rephrased this, hopefully clearer now.